pypeit/coadd3d.py

```diff
@@ -0,0 +1,1853 @@
+            # Compute the weights
+            all_wghts = datacube.compute_weights(self.all_ra, self.all_dec, self.all_wave, self.all_sci, self.all_ivar, self.all_idx, wl_full[:, :, 0],
+                                                      self._dspat, self._dwv, relative_weights=self.cubepar['relative_weights'])
+        return all_wghts
+
+    def coadd(self):
+        """
+        This is the main routine called to convert PypeIt spec2d files into PypeIt DataCube objects. It is specific
+        to the SlicerIFU data.
+
+        First the data are loaded and several corrections are made. These include:
+
+        * A sky frame or model is subtracted from the science data, and the relative spectral illumination
+          of different slices is corrected.
+        * A mask of good pixels is identified
+        * A common spaxel scale is determined, and the astrometric correction is derived
+        * An RA and Dec image is created for each pixel.
+        * Based on atmospheric conditions, a differential atmospheric refraction correction is applied.
+        * Extinction correction
+        * Flux calibration (optional - this calibration is only applied if a standard star cube is supplied)
+
+        If the input frames will not be combined (combine=False) if they won't be aligned (align=False), then
+        each individual spec2d file is converted into a spec3d file (i.e. a PypeIt DataCube object). These fits
+        files can be loaded/viewed in other software packages to display or combine multiple datacubes into a
+        single datacube. However, note that different software packages use combination algorithms that may not
+        conserve flux, or may produce covariance between adjacent voxels.
+
+        If the user wishes to either spatially align multiple exposures (align=True) or combine multiple
+        exposures (combine=True), then the next set of operations include:
+
+        * Generate white light images of each individual cube (according to a user-specified wavelength range)
+        * Align multiple frames if align=True (either manually by user input, or automatically by cross-correlation)
+        * Create the output WCS, and apply the flux calibration to the data
+        * Generate individual datacubes (combine=False) or one master datacube containing all exposures (combine=True).
+          Note, there are several algorithms used to combine multiple frames. Refer to the subpixellate() routine for
+          more details about the combination options.
+        """
+        # No need to continue if we are not combining nor aligning frames
+        if not self.combine and not self.align:
+            return
+
+        # If the user is aligning or combining, the spatial scale of the output cubes needs to be consistent.
+        # Set the spatial and spectral scales of the output datacube
+        self.set_voxel_sampling()
```

Comment:
See my comments above, set_voxel_sampling should return something.  As a general rule,  control flows like 

self.do_x()
self.do_y()
self.do_z()

Are not only inscrutable, but often  do not manage variable scope and are prone to things changing accidentally changing behind the scenes, like if x changes variables that are then required for y. I know that everything is an attribute of the object, but actually that should not really be the case. The fewer global variables the better.